### DISCUSSION TITLE
proposal: Add fieldalignment to generated structs

### DISCUSSION THREAD

**Discussion by switchupcb** [99designs/gqlgen#2070] [Ideas]
# What is fieldalignment?

**Struct padding** aligns fields of a structs to addresses in memory. This is done to improve performance and prevent numerous issues on a system's architecture (32-bit, 64-bit). For a simple explanation, view [Golang Struct Size and Memory optimization (visual example)](https://medium.com/techverito/golang-struct-size-and-memory-optimisation-b46b124f008d
). In Go, fieldalignment can be checked and fixed using the [fieldalignment tool](https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/fieldalignment) which is installed using `go install golang.org/x/tools/go/analysis/passes/fieldalignment/cmd/fieldalignment@latest`. Misaligned fields add more memory-usage to a program, which can effect performance in a numerous amount of ways. **This proposal aims to save memory by adding fieldalignment to gqlgen.**

### Read More
Fieldalignment in Practice: https://itnext.io/structure-size-optimization-in-golang-alignment-padding-more-effective-memory-layout-linters-fffdcba27c61
Memory Layouts: https://go101.org/article/memory-layout.html

## fieldalignment in gqlgen

The following code block contains output from `fieldalignment -fix` on gqlgen structs.
```go
models.gqlgen.go:17:14: struct with 184 pointer bytes could be 176
models.gqlgen.go:31:31: struct with 48 pointer bytes could be 40
models.gqlgen.go:84:23: struct with 64 pointer bytes could be 56
models.gqlgen.go:117:25: struct with 72 pointer bytes could be 56
models.gqlgen.go:148:25: struct with 88 pointer bytes could be 80
models.gqlgen.go:172:22: struct with 40 pointer bytes could be 32
models.gqlgen.go:294:18: struct with 40 pointer bytes could be 32
models.gqlgen.go:327:12: struct with 80 pointer bytes could be 72
models.gqlgen.go:342:16: struct with 56 pointer bytes could be 48
models.gqlgen.go:362:14: struct with 80 pointer bytes could be 64
models.gqlgen.go:377:12: struct with 40 pointer bytes could be 32
models.gqlgen.go:385:11: struct with 48 pointer bytes could be 40
models.gqlgen.go:393:14: struct with 88 pointer bytes could be 80
models.gqlgen.go:409:18: struct with 64 pointer bytes could be 56
models.gqlgen.go:468:23: struct with 80 pointer bytes could be 72
models.gqlgen.go:504:25: struct with 88 pointer bytes could be 72
models.gqlgen.go:518:23: struct with 40 pointer bytes could be 24
models.gqlgen.go:528:22: struct with 48 pointer bytes could be 40
models.gqlgen.go:538:25: struct with 112 pointer bytes could be 104
models.gqlgen.go:575:11: struct with 96 pointer bytes could be 88
```

176 bytes saved. 20 misaligned models. You don't use these models in every request. So for each model used in 1 MILLION requests, _that's..._

**8.8 MB saved per misaligned model**.

Along with better [Garbage Collection](https://agrim123.github.io/posts/go-garbage-collector.html) performance due to a lower amount of peaks.

## Previous Mentions

https://github.com/99designs/gqlgen/issues/1710 (https://github.com/99designs/gqlgen/discussions/1895)

## Steps

- [x] Add code to field align the written file. 
- [ ] Add conditional statement to fieldalign when enabled (unless it's encouraged to always fieldalign).
- [ ] Add documentation on how to enable field alignment and install it using `go install golang.org/x/tools/go/analysis/passes/fieldalignment/cmd/fieldalignment@latest`: **Where should this be configured from in the .yml?**

**Comment by frederikhors** [comment]
As you have already noticed [I proposed this long ago](https://github.com/99designs/gqlgen/issues/1710).
I move this in proposals, we are keeping the issues for problems.

**Comment by frederikhors** [comment]
Please, continue the PR. It will be analyzed and merged ASAP.

**Reply by switchupcb** [reply]
I can't continue the PR. I'm unfamiliar with the rules to place configuration and needed that issue for guidance.

**Reply by switchupcb** [reply]
I can continue if I know where to put it.

**Reply by switchupcb** [reply]
Figured it out, just waiting for approval on the idea.
